### DISCUSSION TITLE
Compilation problems

### DISCUSSION THREAD

**Discussion by jameslkey** [terjeio/ioSender#99] [Q&A]
I am taking a look at the odometer plugin for grblHAL and the possiblity of creating the interface for iosender. I was able to install all of the dependencies except Helixtoolkit. I attempted to use the nuGet packages but can only get a few of them to install and I still get errors when I attempt to compile. Any help would be appreciated.

**Comment by terjeio** [comment]
Which errors? Could it be due to the reference path to the Helix toolkit dll in the _CNC GCodeViewer_ project beeing incorrect for your machine? This since I have not used nuGet to fetch the package?

Here is the content of folder I have referenced:
![image](https://user-images.githubusercontent.com/20260062/117761663-0d822100-b228-11eb-86f3-b8259268a8b3.png)

**Comment by jameslkey** [comment]
As you can see HelixToolkit and the Wpf extension are installed vias nuGet. But HelixToolkit.core.Wpf is not and will not install, no will the other packages that don’t have green check marks.Here are the errors. I did try to compile HelixToolkit from source but was unable to locate all of its dependencies but I am willing to try again.   Sent from Mail for Windows 10 From: Terje IoSent: Tuesday, May 11, 2021 01:14To: terjeio/ioSenderCc: James L. Key; AuthorSubject: Re: [terjeio/ioSender] Compilation problems (#99) Which errors? Could it be due to the reference path to the Helix toolkit dll in the CNC GCodeViewer project beeing incorrect for your machine? This since I have not used nuGet to fetch the package?Here is the content of folder I have referenced:—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by jameslkey** [comment]
Here is the error list when I run build solution on “GCode Sender” Sent from Mail for Windows 10 From: Terje IoSent: Tuesday, May 11, 2021 01:14To: terjeio/ioSenderCc: James L. Key; AuthorSubject: Re: [terjeio/ioSender] Compilation problems (#99) Which errors? Could it be due to the reference path to the Helix toolkit dll in the CNC GCodeViewer project beeing incorrect for your machine? This since I have not used nuGet to fetch the package?Here is the content of folder I have referenced:—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Reply by terjeio** [reply]
Can't see any errors in your comments above...

Here are files I use, can you try with those? Unzip to the same folder as the GCode Sender project is located.

[helix-toolkit.zip](https://github.com/terjeio/ioSender/files/6461626/helix-toolkit.zip)

**Comment by jameslkey** [comment]
I beat my head upon the rock again and seem to have gotten it to compile!!!! Now to look at the grblHAL plugin for the odometer and then figure out enough C# to create a tab to display what we want. Sent from Mail for Windows 10 From: Terje IoSent: Tuesday, May 11, 2021 15:51To: terjeio/ioSenderCc: James L. Key; AuthorSubject: Re: [terjeio/ioSender] Compilation problems (#99) Can't see any errors in your comments above...Here are files I use, can you try with those? Unzip to the same folder as the GCode Sender project is located.helix-toolkit.zip—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by jameslkey** [comment]
I noted that on line 334 of GCodeParser.cs we set isComment if the data from the controller starts with ‘MSG’. The odometer plugin uses ‘MSG’ as it’s preamble. I can imagine we want to keep that in place but maybe catch the [MSG:SPINDLEHRS 4:20][MSG:MOTORHRS 5:52][MSG:ODOMETERX 22.4][MSG:ODOMETERY 19.4][MSG:ODOMETERZ 8.2]separately further down the block?  I guess maybe I should create a pull request and go from there but I’d like to be able to do this with as minimal an impact on the existing code as possible. Is there a struct or other data entity that this is stored in for general program access as opposed to digging in to the ‘core’ of the application?  Sent from Mail for Windows 10

**Comment by terjeio** [comment]
Great!

_GCodeParser.cs_ has nothing to do with this. The responses from the controller is read and parsed by _GrblViewModel.cs_ which sets a number of ViewModel properties. Changes to these can be subscribed to, and the one that may be of interest is `Message`. You can also subscribe to the `OnResponseReceived` event to get the raw data. I use the latter method many places in the code: issue a request to the controller and collect/parse responses until an "ok" response is received. The `GrblInfo` class is in _Grbl.cs_ is a good example - see the `Get()` method(s).

The only change that is needed in the existing code is adding a dialog, a tab is a bit over the top? A dialog can be shown via a menu/menu item. When the dialog is shown, request the data and display them. _About.xaml_ in the CNC Controls project does that and uses the `Get()` method mentioned above to fetch the data. You have to add your own get method for the odometer data though.

If you are not familiar with WPF and the [VMMV programming paradigm](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93viewmodel) you might want to read up a bit on those.
